# HG changeset patch
# User David Bateman <dbateman@free.fr>
# Date 1289197990 -3600
#      Mon Nov 08 07:33:10 2010 +0100
# Node ID d2f3a904ac6b91286aa01164f56bfda1c21bcc19
# Parent  d468f5c1095595ac89261190373a7e7910a92fb6
Unset the foreground color for the gnuplot backend earlier (fixes #31533)

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,13 @@
+2010-11-08  David Bateman  <dbateman@free.fr>
+
+	* plot/__go_draw_axes__.m: Remove foreground color earlier.
+	* plot/__go_draw_figure__.m: Don't set the foreground color here.
+
 2010-11-08  Rik  <octave@nomad.inbox5.com>
 
 	* sparse/sprand.m: Add input validation to function.
 
 2010-11-08  Ben Abbott <bpabbott@mac.com>
 
 	* plot/private/__bar__.m: Permit bar() to be called with scalar inputs.
 	* plot/__go_draw_axes__.m, plot/axis.m: Set proper tight axis limits
diff --git a/scripts/plot/__go_draw_axes__.m b/scripts/plot/__go_draw_axes__.m
--- a/scripts/plot/__go_draw_axes__.m
+++ b/scripts/plot/__go_draw_axes__.m
@@ -18,19 +18,20 @@
 
 ## -*- texinfo -*-
 ## @deftypefn {Function File} {} __go_draw_axes__ (@var{h}, @var{plot_stream}, @var{enhanced}, @var{mono})
 ## Undocumented internal function.
 ## @end deftypefn
 
 ## Author: jwe
 
-function __go_draw_axes__ (h, plot_stream, enhanced, mono, bg_is_set, hlgnd)
+function __go_draw_axes__ (h, plot_stream, enhanced, mono, 
+                           bg_is_set, fg_is_set, hlgnd)
 
-  if (nargin >= 4 && nargin <= 6)
+  if (nargin >= 4 && nargin <= 7)
 
     showhiddenhandles = get (0, "showhiddenhandles");
     unwind_protect
       set (0, "showhiddenhandles", "on");
       axis_obj = __get__ (h);
     unwind_protect_cleanup
       set (0, "showhiddenhandles", showhiddenhandles);
     end_unwind_protect
@@ -1509,24 +1510,32 @@ function __go_draw_axes__ (h, plot_strea
                    usingclause{i}, titlespec{i}, withclause{i});
         elseif (is_image_data (i))
           if (! is_image_data (i-1))
             fputs (plot_stream, "; ");
             if (bg_is_set)      
               fputs (plot_stream, "unset obj 1; \\\n");
               bg_is_set = false;
             endif
+            if (fg_is_set)
+              fputs (plot_stream, "unset obj 2; \\\n");
+              fg_is_set = false;
+            endif
           endif
           fprintf (plot_stream, "%s \"-\" %s %s %s \\\n", plot_cmd,
                    usingclause{i}, titlespec{i}, withclause{i});
         elseif (is_image_data (i-1))
           if (bg_is_set)      
             fputs (plot_stream, "unset obj 1; \\\n");
-              bg_is_set = false;
-            endif
+            bg_is_set = false;
+          endif
+          if (fg_is_set)
+            fputs (plot_stream, "unset obj 2; \\\n");
+            fg_is_set = false;
+          endif
           fprintf (plot_stream, "%s \"-\" binary format='%%float64' %s %s %s \\\n", plot_cmd,
                    usingclause{i}, titlespec{i}, withclause{i});
         else
           fprintf (plot_stream, ", \"-\" binary format='%%float64' %s %s %s \\\n",
                    usingclause{i}, titlespec{i}, withclause{i});
         endif
       endfor
       fputs (plot_stream, ";\n");
diff --git a/scripts/plot/__go_draw_figure__.m b/scripts/plot/__go_draw_figure__.m
--- a/scripts/plot/__go_draw_figure__.m
+++ b/scripts/plot/__go_draw_figure__.m
@@ -112,17 +112,17 @@ function __go_draw_figure__ (h, plot_str
                       elseif (! isempty (lm))
                         toberm = [toberm, line("xdata",[0,0],"ydata",[0,0], "color", get(lm,"color"), "linestyle", "none", "marker", get(lm,"marker"), "markeredgecolor", get(lm,"markeredgecolor"), "markerfacecolor", get(lm,"markerfacecolor"), "markersize", get (lm, "markersize"), "displayname", displayname, "parent", kids(i))];
                       endif
                     endfor
                     if (bg_is_set)
                       fprintf (plot_stream, "set border linecolor rgb \"#%02x%02x%02x\"\n", 255 * (1 - bg));
                     endif
                     __go_draw_axes__ (kids(i), plot_stream, enhanced, mono,
-                                      bg_is_set, hlgnd);
+                                      bg_is_set, false, hlgnd);
                   unwind_protect_cleanup
                     ## Return axes "units" and "position" back to
                     ## their original values.
                     set (ca, "units", orig_axes_units);
                     set (kids(i), "units", legend_axes_units, 
                          "box", legend_axes_box, 
                          "ylim", legend_axes_ylim, 
                          "position", legend_axes_position,
@@ -160,26 +160,24 @@ function __go_draw_figure__ (h, plot_str
                       if (isscalar(udata.handle) 
                           && ! isempty (intersect (udata.handle, kids (i))))
                         hlegend = get (fkids (j));
                         break;
                       endif
                     endif
                   endfor
                   __go_draw_axes__ (kids(i), plot_stream, enhanced, mono,
-                                    bg_is_set, hlegend);
+                                    bg_is_set, fg_is_set, hlegend);
                 unwind_protect_cleanup
                   ## Return axes "units" and "position" back to
                   ## their original values.
                   set (kids(i), "units", orig_axes_units);
                   set (kids(i), "position", orig_axes_position);
                   bg_is_set = false;
-                  if (fg_is_set)
-                    fputs (plot_stream, "unset obj 2\n");
-                  endif
+                  fg_is_set = false;
                 end_unwind_protect
               endif
             case "uimenu"
               ## ignore uimenu objects
             otherwise
               error ("__go_draw_figure__: unknown object class, %s", type);
           endswitch
         endfor
